### PR TITLE
[CARBONDATA-3291] Fix that MV datamap doesn't take affect when the same table join

### PR DESCRIPTION
**Pull request by qiuchenjian**
[Problem]
MV datamap doesn't take affect when the same table join
the error scene see the test case

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2569/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10828/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2799/

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/Navigator.scala [hunk 1]
```diff
@@ -146,21 +146,28 @@ private[mv] class Navigator(catalog: SummaryDatasetCatalog, session: MVSession)
...
-        } else {
-          pair._2
+          pair._1 match {
+            case relation: HarmonizedRelation if relation.hasTag =>
+              pair._2.asInstanceOf[HarmonizedRelation].addTag
+            case _ =>
+              pair._2
+          }
+        val nxtSubsumer = curSubsumer.transform {
+          case node: ModularRelation if node.asInstanceOf[ModularRelation].
```
Comment (ravipesala) [inline datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/Navigator.scala]: No need to typecast here , you can directly write `f node.fineEquals(pair._1)`

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/Navigator.scala [hunk 2]
```diff
@@ -171,4 +178,27 @@ private[mv] class Navigator(catalog: SummaryDatasetCatalog, session: MVSession)
...
+                          rIndex: Int, eIndex: Int): Boolean = {
+    (rtable, etable) match {
+      case _: (ModularRelation, ModularRelation) =>
+        val rtableParent = subsumer.find(p => p.children.contains(rtable)).get
+        val etableParent = subsumee.find(p => p.children.contains(etable)).get
+        (rtableParent, etableParent) match {
+          case _: (Select, Select) =>
+            val rJoinEdges = rtableParent.asInstanceOf[Select].joinEdges
+            val eJoinEdges = etableParent.asInstanceOf[Select].joinEdges
+            val intersetJoinEdges: Seq[JoinEdge] = rJoinEdges intersect eJoinEdges
```
Comment (ravipesala) [inline datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/Navigator.scala]: Can be simplified as below
```
case (e: Select, r: Select) =>
            val intersetJoinEdges = r.joinEdges intersect e.joinEdges
```

**Comment by qiuchenjian** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/3169/

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/Navigator.scala (hunk 2 above)
Comment (qiuchenjian) [reply datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/Navigator.scala]: done

File: datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/Navigator.scala (hunk 1 above)
Comment (qiuchenjian) [reply datamap/mv/core/src/main/scala/org/apache/carbondata/mv/rewrite/Navigator.scala]: done

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/11434/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/3366/

**Comment by ravipesala** [comment]
LGTM

**Pull request closed without merging**